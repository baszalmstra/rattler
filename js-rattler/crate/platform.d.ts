/**
 * All platform names supported by this library.
 *
 * @public
 */
export declare const platformNames: readonly [
    "noarch",
    "linux-32",
    "linux-64",
    "linux-aarch64",
    "linux-armv6l",
    "linux-armv7l",
    "linux-ppc64le",
    "linux-ppc64",
    "linux-ppc",
    "linux-s390x",
    "linux-riscv32",
    "linux-riscv64",
    "osx-64",
    "osx-arm64",
    "win-32",
    "win-64",
    "win-arm64",
    "emscripten-wasm32",
    "wasi-wasm32",
    "zos-z",
];

/**
 * A type that represents a valid platform.
 *
 * @public
 */
export declare type Platform = (typeof platformNames)[number];

/**
 * A type that represents a valid architecture.
 *
 * @public
 */
export declare type Arch = (typeof archNames)[number];

/**
 * All architecture names supported by this library.
 *
 * @public
 */
export declare const archNames: readonly [
    "x86",
    "x86_64",
    "aarch64",
    "arm64",
    "armv6l",
    "armv7l",
    "ppc64le",
    "ppc64",
    "ppc",
    "s390x",
    "riscv32",
    "riscv64",
    "wasm32",
    "z",
];
