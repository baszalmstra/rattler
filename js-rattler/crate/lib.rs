mod error;
mod gateway;
pub mod solve;
mod utils;
mod version;
mod version_spec;
mod parse_strictness;

pub use error::{JsError, JsResult};

use wasm_bindgen::prelude::*;

// When the `wee_alloc` feature is enabled, use `wee_alloc` as the global
// allocator.
#[cfg(feature = "wee_alloc")]
#[global_allocator]
static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;

/// This function is called when the wasm module is instantiated.
#[wasm_bindgen(start)]
pub fn start() {
    utils::set_panic_hook();
}
