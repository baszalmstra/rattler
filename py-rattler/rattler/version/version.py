from __future__ import annotations

from typing import Optional, Tuple

from rattler.rattler import PyVersion


class Version:
    def __init__(self, version: str):
        if isinstance(version, str):
            self._version = PyVersion(version)
        else:
            raise TypeError(
                "Version constructor received unsupported type "
                f" {type(version).__name__!r} for the `version` parameter"
            )

    @classmethod
    def _from_py_version(cls, py_version: PyVersion) -> Version:
        """Construct Rattler version from FFI PyVersion object."""
        version = cls.__new__(cls)
        version._version = py_version
        return version

    def __str__(self) -> str:
        return self._version.as_str()

    def __repr__(self) -> str:
        return self.__str__()

    @property
    def epoch(self) -> Optional[str]:
        """
        Gets the epoch of the version or `None` if the epoch was not defined.

        Examples
        --------
        >>> v = Version('2!1.0')
        >>> v.epoch
        2
        """
        return self._version.epoch()

    def bump(self) -> Version:
        """
        Returns a new version where the last numerical segment of this version has
        been bumped.

        Examples
        --------
        >>> v = Version('1.0')
        >>> v.bump()
        1.1
        """
        return Version._from_py_version(self._version.bump())

    @property
    def local(self) -> bool:
        """
        Returns true if this version has a local segment defined.

        Examples
        --------
        >>> v = Version('1.0')
        >>> v.local
        False
        """
        return self._version.has_local()

    def as_major_minor(self) -> Optional[Tuple[int, int]]:
        """
        Returns the major and minor segments from the version.

        Examples
        --------
        >>> v = Version('1.0')
        >>> v.as_major_minor()
        (1, 0)
        """
        return self._version.as_major_minor()

    @property
    def dev(self) -> bool:
        """
        Returns true if the version contains a component name "dev".

        Examples
        --------
        >>> v = Version('1.0.1dev')
        >>> v.dev
        True
        """
        return self._version.is_dev()

    def starts_with(self, other: Version) -> bool:
        """
        Checks if the version and local segment start
        same as other version.

        Examples
        --------
        >>> v1 = Version('1.0.1')
        >>> v2 = Version('1.0')
        >>> v1.starts_with(v2)
        True
        """
        return self._version.starts_with(other._version)

    def compatible_with(self, other: Version) -> bool:
        """
        Checks if this version is compatible with other version.
        """
        return self._version.compatible_with(other._version)

    def pop_segments(self, n: int = 1) -> Optional[Version]:
        """
        Pops `n` number of segments from the version and returns
        the new version. Returns `None` if the version becomes
        invalid due to the operation.
        """
        new_py_version = self._version.pop_segments(n)
        if new_py_version:
            # maybe it should raise an exception instead?
            return self._from_py_version(new_py_version)

    def with_segments(self, start: int, stop: int) -> Optional[Version]:
        """
        Returns new version with with segments ranging from `start` to `stop`.
        `stop` is exclusive.
        """
        new_py_version = self._version.with_segments(start, stop)
        if new_py_version:
            return self._from_py_version(new_py_version)
        else:
            # maybe it should raise an exception instead?
            return None

    def segment_count(self) -> int:
        """
        Returns the number of segments in the version.
        """
        return self._version.segment_count()

    def strip_local(self) -> Version:
        """
        Returns a new version with local segment stripped.
        """
        return self._from_py_version(self._version.strip_local())

    def __eq__(self, other: Version) -> bool:
        return self._version.equals(other._version)

    def __ne__(self, other: Version) -> bool:
        return self._version.not_equal(other._version)

    def __gt__(self, other: Version) -> bool:
        return self._version.greater_than(other._version)

    def __lt__(self, other: Version) -> bool:
        return self._version.less_than(other._version)

    def __ge__(self, other: Version) -> bool:
        return self._version.greater_than_equals(other._version)

    def __le__(self, other: Version) -> bool:
        return self._version.less_than_equals(other._version)
