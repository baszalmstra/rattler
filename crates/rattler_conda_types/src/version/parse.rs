use super::{Component, Version};
use crate::version::{LOCAL_VERSION_MASK, LOCAL_VERSION_OFFSET};
use smallvec::SmallVec;
use std::{
    convert::Into,
    default::Default,
    error::Error,
    fmt::{Display, Formatter},
    num::ParseIntError,
    result::Result,
    str::FromStr,
};

/// An error that occurred during parsing of a string to a version.
#[derive(Debug, Clone, Eq, PartialEq)]
pub struct ParseVersionError {
    /// The original string that was the input of the parser
    pub version: String,

    /// The type of parse error that occurred
    pub kind: ParseVersionErrorKind,
}

impl Display for ParseVersionError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "malformed version string '{}': ", &self.version)?;
        match &self.kind {
            ParseVersionErrorKind::Empty => write!(f, "empty string"),
            ParseVersionErrorKind::InvalidCharacters => write!(f, "invalid character(s)"),
            ParseVersionErrorKind::EpochMustBeInteger(e) => {
                write!(f, "epoch must be an integer: {}", e)
            }
            ParseVersionErrorKind::DuplicateEpochSeparator => {
                write!(f, "duplicated epoch separator '!'")
            }
            ParseVersionErrorKind::DuplicateLocalVersionSeparator => {
                write!(f, "duplicated local version separator '+'")
            }
            ParseVersionErrorKind::EmptyVersionComponent => write!(f, "empty version component"),
            ParseVersionErrorKind::InvalidNumeral(e) => write!(f, "invalid numeral: {}", e),
            ParseVersionErrorKind::TooManySegments => write!(f, "too many segments"),
            ParseVersionErrorKind::TooManyComponentsInASegment => write!(f, "too many version components, a single version segment can at most contain {} components", (1<<16)-1),
        }
    }
}

impl Error for ParseVersionError {}

impl ParseVersionError {
    /// Create a new parse error
    pub fn new(text: impl Into<String>, kind: ParseVersionErrorKind) -> Self {
        Self {
            version: text.into(),
            kind,
        }
    }
}

/// The type of parse error that occurred when parsing a version string.
#[derive(Debug, Eq, PartialEq, Clone)]
pub enum ParseVersionErrorKind {
    /// The string was empty
    Empty,
    /// The string contained invalid characters
    InvalidCharacters,
    /// The epoch was not an integer value
    EpochMustBeInteger(ParseIntError),
    /// The string contained an invalid numeral
    InvalidNumeral(ParseIntError),
    /// The string contained multiple epoch separators
    DuplicateEpochSeparator,
    /// The string contained multiple local version separators
    DuplicateLocalVersionSeparator,
    /// The string contained an empty version component
    EmptyVersionComponent,
    /// Too many segments.
    TooManySegments,
    /// Too many segments.
    TooManyComponentsInASegment,
}

/// Returns true if the specified char is a valid char for a version string.
pub(crate) fn is_valid_char(c: char) -> bool {
    matches!(c, '.'|'+'|'!'|'_'|'0'..='9'|'a'..='z')
}

/// Returns true if the specified string contains only valid chars for a version string.
fn has_valid_chars(version: &str) -> bool {
    version.chars().all(is_valid_char)
}

impl FromStr for Version {
    type Err = ParseVersionError;

    // Implementation taken from https://github.com/conda/conda/blob/0050c514887e6cbbc1774503915b45e8de12e405/conda/models/version.py#L47

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        // Version comparison is case-insensitive so normalize everything to lowercase
        let normalized = s.trim().to_lowercase();

        // Basic validity check
        if normalized.is_empty() {
            return Err(ParseVersionError::new(s, ParseVersionErrorKind::Empty));
        }

        // Allow for dashes as long as there are no underscores as well. Dashes are then converted
        // to underscores.
        let lowered = if normalized.contains('-') && !normalized.contains('_') {
            normalized.replace('-', "_")
        } else {
            normalized
        };

        // Ensure the string only contains valid characters
        if !has_valid_chars(&lowered) {
            return Err(ParseVersionError::new(
                s,
                ParseVersionErrorKind::InvalidCharacters,
            ));
        }

        // Find epoch
        let (epoch, rest) = if let Some((epoch, rest)) = lowered.split_once('!') {
            let epoch: u64 = epoch.parse().map_err(|e| {
                ParseVersionError::new(s, ParseVersionErrorKind::EpochMustBeInteger(e))
            })?;
            (Some(epoch), rest)
        } else {
            (None, lowered.as_str())
        };

        // Ensure the rest of the string no longer contains an epoch
        if rest.find('!').is_some() {
            return Err(ParseVersionError::new(
                s,
                ParseVersionErrorKind::DuplicateEpochSeparator,
            ));
        }

        // Find local version string
        let (local, rest) = if let Some((rest, local)) = rest.rsplit_once('+') {
            (local, rest)
        } else {
            ("", rest)
        };

        // Ensure the rest of the string no longer contains a local version separator
        if rest.find('+').is_some() {
            return Err(ParseVersionError::new(
                s,
                ParseVersionErrorKind::DuplicateLocalVersionSeparator,
            ));
        }

        // Split the local version by '_' or '.'
        let local_split = local.split(&['.', '_'][..]);

        // If the last character of a version is '-' or '_', don't split that out individually.
        // Implements the instructions for openssl-like versions. You can work-around this problem
        // by appending a dash to plain version numbers.
        let version: SmallVec<[String; 6]> = if rest.ends_with('_') {
            let mut versions: SmallVec<[String; 6]> = rest[..(rest.len() as isize - 1) as usize]
                .replace('_', ".")
                .split('.')
                .map(ToOwned::to_owned)
                .collect();
            if let Some(last) = versions.last_mut() {
                *last += "_";
            }
            versions
        } else {
            rest.replace('_', ".")
                .split('.')
                .map(ToOwned::to_owned)
                .collect()
        };
        let version_split = version.iter().map(|s| s.as_str());

        let mut components = SmallVec::default();
        let mut segments = SmallVec::default();
        let mut flags = 0u8;

        if let Some(epoch) = epoch {
            components.push(epoch.into());
            flags |= 0x1; // Mark that the version contains an epoch
        }

        fn split_component<'a>(
            segments_iter: impl Iterator<Item = &'a str>,
            segments: &mut SmallVec<[u16; 4]>,
            components: &mut SmallVec<[Component; 3]>,
        ) -> Result<(), ParseVersionErrorKind> {
            for component in segments_iter {
                let version_split_re = lazy_regex::regex!(r#"([0-9]+|[^0-9]+)"#);
                let mut numeral_or_alpha_split = version_split_re.find_iter(component).peekable();
                if numeral_or_alpha_split.peek().is_none() {
                    return Err(ParseVersionErrorKind::EmptyVersionComponent);
                }

                let mut atoms = numeral_or_alpha_split
                    .map(|mtch| match mtch.as_str() {
                        num if num.chars().all(|c| c.is_ascii_digit()) => num
                            .parse()
                            .map_err(ParseVersionErrorKind::InvalidNumeral)
                            .map(Component::Numeral),
                        "post" => Ok(Component::Post),
                        "dev" => Ok(Component::Dev),
                        ident => Ok(Component::Iden(ident.to_owned().into_boxed_str())),
                    })
                    .peekable();

                // A segment must always starts with a numeral
                let mut component_count = 0u16;
                if !matches!(atoms.peek(), Some(&Ok(Component::Numeral(_)))) {
                    components.push(Component::Numeral(0));
                    component_count = component_count
                        .checked_add(1)
                        .ok_or(ParseVersionErrorKind::TooManyComponentsInASegment)?;
                }

                // Add the components
                for component in atoms {
                    components.push(component?);
                    component_count = component_count
                        .checked_add(1)
                        .ok_or(ParseVersionErrorKind::TooManyComponentsInASegment)?;
                }

                // Add the segment information
                segments.push(component_count);
            }

            Ok(())
        }

        split_component(version_split, &mut segments, &mut components)
            .map_err(|e| ParseVersionError::new(s, e))?;

        if !local.is_empty() {
            if segments.len() > (LOCAL_VERSION_MASK >> LOCAL_VERSION_OFFSET) as usize {
                // There are too many segments to be able to encode the local segment parts into the
                // special `flag` we store. The flags is 8 bits and the first bit is used to
                // indicate if there is an epoch or not. The remaining 7 bits are used to indicate
                // which segment is the first that belongs to the local version part. We can encode
                // at most 127 positions so if there are more segments in the common version part,
                // we cannot represent this version.
                return Err(ParseVersionError::new(
                    s,
                    ParseVersionErrorKind::TooManySegments,
                ));
            }

            // Encode that the local version segment starts at the given index.
            flags |= (u8::try_from(segments.len()).unwrap()) << LOCAL_VERSION_OFFSET;

            split_component(local_split, &mut segments, &mut components)
                .map_err(|e| ParseVersionError::new(s, e))?
        };

        Ok(Self {
            norm: Some(lowered.into_boxed_str()),
            flags,
            segment_lengths: segments,
            components,
        })
    }
}
