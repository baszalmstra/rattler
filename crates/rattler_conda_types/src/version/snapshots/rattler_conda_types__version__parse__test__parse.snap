---
source: crates/rattler_conda_types/src/version/parse.rs
assertion_line: 607
expression: index_map
---
{
    "$": Error(
        "expected a version component e.g. `2` or `rc`",
    ),
    ".": Error(
        "expected a version component e.g. `2` or `rc`",
    ),
    "1!1.2a.3-rc1": Version(
        Version {
            norm: None,
            epoch: Some(
                1,
            ),
            version: [[1], [2, a], [3], [0, rc, 1]],
            local: [],
        },
    ),
    "1+": Error(
        "expected a version component e.g. `2` or `rc`",
    ),
    "1+$": Error(
        "expected a version component e.g. `2` or `rc`",
    ),
    "1+.": Error(
        "expected a version component e.g. `2` or `rc`",
    ),
    "1+2": Version(
        Version {
            norm: None,
            epoch: None,
            version: [[1]],
            local: [[2]],
        },
    ),
    "1-2-3": Version(
        Version {
            norm: None,
            epoch: None,
            version: [[1], [2], [3]],
            local: [],
        },
    ),
    "1-2-3_": Version(
        Version {
            norm: None,
            epoch: None,
            version: [[1], [2], [3, _]],
            local: [],
        },
    ),
    "1-2_3": Error(
        "cannot use both underscores and dashes as version segment seperators",
    ),
    "1.0.1_": Version(
        Version {
            norm: None,
            epoch: None,
            version: [[1], [0], [1, _]],
            local: [],
        },
    ),
    "1.0.1post.za": Version(
        Version {
            norm: None,
            epoch: None,
            version: [[1], [0], [1, post], [0, za]],
            local: [],
        },
    ),
    "1@2": Error(
        "expected a '.', '-', or '_'",
    ),
    "1_": Version(
        Version {
            norm: None,
            epoch: None,
            version: [[1, _]],
            local: [],
        },
    ),
    "1_2_3": Version(
        Version {
            norm: None,
            epoch: None,
            version: [[1], [2], [3]],
            local: [],
        },
    ),
    "1_2_3_": Version(
        Version {
            norm: None,
            epoch: None,
            version: [[1], [2], [3, _]],
            local: [],
        },
    ),
    "1__": Version(
        Version {
            norm: None,
            epoch: None,
            version: [[1], [0, _]],
            local: [],
        },
    ),
    "1___": Error(
        "expected a version component e.g. `2` or `rc`",
    ),
}
